vernac/record.ml

```diff
@@ -291,26 +326,97 @@ let instantiate_possibly_recursive_type ind u ntypes paramdecls fields =
   Termops.substl_rel_context (subst @ subst') fields
 
-type projection_flags = {
-  pf_subclass: bool;
-  pf_canonical: bool;
-}
-
 (* We build projections *)
-let declare_projections indsp ctx ?(kind=Decls.StructureComponent) binder_name flags fieldimpls fields =
+
+(* TODO: refactor the declaration part here; this requires some
+   surgery as Evarutil.finalize is called too early in the path *)
+let build_named_proj ~primitive ~flags ~poly ~univs ~uinstance ~kind env paramdecls paramargs decl impls fid subst sp_projs nfi ti i indsp mib lifted_fields x rp =
+  let ccl = subst_projection fid subst ti in
+  let body, p_opt = match decl with
+    | LocalDef (_,ci,_) -> subst_projection fid subst ci, None
+    | LocalAssum ({binder_relevance=rci},_) ->
+      (* [ccl] is defined in context [params;x:rp] *)
+      (* [ccl'] is defined in context [params;x:rp;x:rp] *)
+      if primitive then
+        let p = Projection.Repr.make indsp
+            ~proj_npars:mib.mind_nparams ~proj_arg:i (Label.of_id fid) in
+        mkProj (Projection.make p true, mkRel 1), Some p
+      else
+        let ccl' = liftn 1 2 ccl in
+        let p = mkLambda (x, lift 1 rp, ccl') in
+        let branch = it_mkLambda_or_LetIn (mkRel nfi) lifted_fields in
+        let ci = Inductiveops.make_case_info env indsp rci LetStyle in
+        (* Record projections are always NoInvert because they're at
+           constant relevance *)
+        mkCase (ci, p, NoInvert, mkRel 1, [|branch|]), None
+  in
+  let proj = it_mkLambda_or_LetIn (mkLambda (x,rp,body)) paramdecls in
+  let projtyp = it_mkProd_or_LetIn (mkProd (x,rp,ccl)) paramdecls in
+  let entry = Declare.definition_entry ~univs ~types:projtyp proj in
+  let kind = Decls.IsDefinition kind in
+  let kn =
+    try Declare.declare_constant ~name:fid ~kind (Declare.DefinitionEntry entry)
+    with Type_errors.TypeError (ctx,te) as exn when not primitive ->
+      let _, info = Exninfo.capture exn in
+      Exninfo.iraise (NotDefinable (BadTypedProj (fid,ctx,te)),info)
+  in
+  Declare.definition_message fid;
+  let term = match p_opt with
+    | Some p ->
+      let _ = DeclareInd.declare_primitive_projection p kn in
+      mkProj (Projection.make p false,mkRel 1)
+    | None ->
+      let proj_args = (*Rel 1 refers to "x"*) paramargs@[mkRel 1] in
+      match decl with
+      | LocalDef (_,ci,_) when primitive -> body
+      | _ -> applist (mkConstU (kn,uinstance),proj_args)
+  in
+  let refi = GlobRef.ConstRef kn in
+  Impargs.maybe_declare_manual_implicits false refi impls;
+  if flags.pf_subclass then begin
+    let cl = ComCoercion.class_of_global (GlobRef.IndRef indsp) in
+    ComCoercion.try_add_new_coercion_with_source refi ~local:false ~poly ~source:cl
+  end;
+  let i = if is_local_assum decl then i+1 else i in
+  (Some kn::sp_projs, i, Projection term::subst)
+
+let build_proj env mib indsp primitive x rp lifted_fields ~poly paramdecls paramargs ~uinstance ~kind ~univs
```

Comment:
Can you write a comment documenting what this function does?

Reply:
Done, it is pretty obvious tho IIANM.